### PR TITLE
Add: Node SDK Tests

### PR DESCRIPTION
**Pull request by akshat1423**
## Description

This PR adds a comprehensive suite of integration and advanced tests for the Mem0 Node SDK, covering both the OSS TypeScript implementation and the client API. The tests ensure robust coverage of core memory operations, vector store integrations, and advanced features.  
This improves reliability, developer confidence, and future maintainability.

#### ✅ `memory.basic.test.ts`
- **Purpose:**  
  Covers CRUD operations for the `Memory` class:  
  `add`, `get`, `update`, `delete`, `getAll`, `search`, `reset`
- **Details:**  
  Uses OpenAI embedder with in-memory vector store  
  Handles API key setup and test timeouts

#### 🕒 `memory.advanced.test.ts` *(commented out)*
- **Planned Scenarios:**  
  Batch operations, multi-store support, error handling

---

#### ✅ `memoryClient.test.ts`
- **Purpose:**  
  Tests simple `MemoryClient` operations:  
  `add`, `get`, `update`, `delete`, `list`, `search`
- **Details:**  
  Uses real API calls with scoped org/project context

#### ✅ `memoryClient.integration.test.ts`
- **Purpose:**  
  End-to-end tests for the `MemoryClient` API
- **Coverage:**  
  - Organization and project creation  
  - Memory CRUD, batch update/delete  
  - Feedback, user listing  
  - Validates org/project context and uses shared utilities

#### ✅ `memoryClient.advanced.test.ts`
- **Purpose:**  
  Covers advanced features and edge cases
- **Coverage:**  
  - Metadata and filter-based search  
  - Agent/user memory separation  
  - Async mode, pagination  
  - Bulk operations: `batchUpdate`, `batchDelete`, `deleteAll`

#### ✅ `memoryClient.typecheck.test.ts`
- **Purpose:**  
  Ensures type correctness of all `MemoryClient` APIs
- **Coverage:**  
  - Method input/output validation  
  - Compile-time checks for misuse of filters, metadata, options  
  - Guards against incorrect org/project typing
- **Note:**  
  Uses tools like [`tsd`](https://github.com/SamVerschueren/tsd) or `expect-type`  
  These tests are verified by the TypeScript compiler and do not run at runtime

**Motivation & Context:**  
- Ensure all core and advanced memory operations are tested.
- Provide coverage for multiple vector store backends and API features.
- Facilitate easier debugging and onboarding for contributors.

**Dependencies:**  
- Requires valid OpenAI and/or Mem0 API keys in your environment or `.env` file.
- Uses Jest as the test runner.

Fixes # (issue)

---

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (does not change functionality, e.g. code style improvements, linting)
- [ ] Documentation update

---

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Unit Test
- [x] Test Script (please provide)

**Instructions to Reproduce:**

1. Set up your `.env` file with the required API keys:
   - For OSS tests: `OPENAI_API_KEY`
   - For client tests: `MEM0_API_KEY` (and optionally `MEM0_API_HOST`)
2. Run tests from the `mem0-ts` directory:
   ```bash
   npm test
   ```
3. All tests should pass if the environment is correctly configured.

**Test Coverage Includes:**
- OSS Memory: add, get, update, delete, getAll, search, reset
- Client SDK: org/project creation, CRUD, batch ops, feedback, user listing, advanced filters, async mode, pagination

---

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

---

## Maintainer Checklist

- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] Made sure Checks passed

---

### CODE REVIEW

File: mem0-ts/src/client/tests/memoryClient.advanced.test.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+  beforeAll(() => {
+    if (!apiKey) throw new Error("MEM0_API_KEY is required in .env");
+    client = new MemoryClient({ apiKey, host });
+  });
+
+  it("adds a memory with metadata and retrieves it by metadata filter", async () => {
+    const userId = randomString();
+    const messages = [
+      { role: "user" as const, content: "I love vegan food." },
```
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.advanced.test.ts:30]: Any particular reason why we need to typecast this to a const?

File: mem0-ts/src/client/tests/memoryClient.integration.test.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    const orgName = 'Test Organization ' + Date.now();
+    const orgRes = await createOrganization(apiKey, orgName);
+    expect(typeof orgRes.org_id).toBe("string");
+    // Do not use this org for other tests
+  });
+
+  it("creates a project in the organization", async () => {
+    // Use the first org for project creation test, but do not use this project for other tests
+    const projectName = 'Test Project ' + Date.now();
+    const projRes = await createProject(apiKey, orgId, projectName);
```
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.integration.test.ts]: We don't need these APIs to test the APIs, this is supposed to be a tests for the SDK

File: mem0-ts/src/client/tests/memoryClient.integration.test.ts [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+function randomString() {
+  return (
+    Math.random().toString(36).substring(2, 15) +
+    Math.random().toString(36).substring(2, 15)
+  );
+}
+
+// Utility to create an organization
+async function createOrganization(apiKey, orgName) {
+  const response = await fetch('https://api.mem0.ai/api/v1/orgs/organizations/', {
```
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.integration.test.ts]: Why do we even need to create a new org for each test?

File: mem0-ts/src/client/tests/memoryClient.integration.test.ts [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+  });
+
+  it("lists users", async () => {
+    const users = await client.users();
+    expect(users).toHaveProperty("results");
+    expect(Array.isArray(users.results)).toBe(true);
+  });
+
+  it("gets organizations", async () => {
+    const orgs = await getOrganizations(apiKey);
```
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.integration.test.ts]: Same as above

File: mem0-ts/src/client/tests/memoryClient.integration.test.ts
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.integration.test.ts:1]: This file in general tests the APIs and not particularly the SDK, can you please share your reasoning behind this choice?

File: mem0-ts/src/client/tests/memoryClient.typecheck.test.ts
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.typecheck.test.ts:1]: I believe this is still a WIP, we need more types check for different cases and it should match our docs.

**Comment by deshraj** [comment]
Can we make sure that we run this as part of the CI?

File: mem0-ts/src/client/tests/memoryClient.advanced.test.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  it("adds a memory for an agent and retrieves it", async () => {
+    const agentId = randomString();
+    const messages = [
+      { role: "assistant" as const, content: "My name is Alice. I am an AI tutor." }
+    ];
+    await client.add(messages, { agent_id: agentId });
+
+    const results = await client.search("Alice", { agent_id: agentId });
+    expect(results.length).toBeGreaterThan(0);
+    expect(results[0].memory).toContain("Alice");
```
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.advanced.test.ts:52]: What if the memory AI tutor appears first? Then this test can fail without a reason. We're on the mercy of the LLMs here.

File: mem0-ts/src/client/tests/memoryClient.advanced.test.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  it("adds a memory for both user and agent and retrieves for each", async () => {
+    const userId = randomString();
+    const agentId = randomString();
+    const messages = [
+      { role: "user" as const, content: "I'm travelling to San Francisco" },
+      { role: "assistant" as const, content: "I'm going to Dubai next month." }
+    ];
+    await client.add(messages, { user_id: userId, agent_id: agentId });
+
+    const userResults = await client.search("San Francisco", { user_id: userId });
```
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.advanced.test.ts:64]: I have seen we're doing most test using search V1, which is alright but it's already deprecated so we should move to V2

File: mem0-ts/src/client/tests/memoryClient.advanced.test.ts [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+      metadata: { food: "vegan" }
+    });
+
+    const filters = {
+      AND: [
+        { metadata: { food: "vegan" } }
+      ]
+    };
+    const results = await client.search("vegan", {
+      user_id: userId,
```
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.advanced.test.ts:98]: Does this param work with API V2?

File: mem0-ts/src/client/tests/memoryClient.advanced.test.ts [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+      api_version: "v2",
+      filters
+    });
+    expect(results.length).toBeGreaterThan(0);
+  });
+
+  it("retrieves paginated memories", async () => {
+    const userId = randomString();
+    for (let i = 0; i < 5; i++) {
+      await client.add([{ role: "user" as const, content: `Memory ${i}` }], { user_id: userId });
```
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.advanced.test.ts:108]: This shouldn't work either, because here I don't see infer false and hence I don't see there's any information to be extracted. If it worked mercy of the LLM

File: mem0-ts/src/client/tests/memoryClient.integration.test.ts [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    const memoryId = res[0].id;
+    // batchDelete expects [memoryId]
+    const deleteRes = await client.batchDelete([memoryId]);
+    expect(typeof deleteRes).toBe("object");
+    expect(deleteRes).toHaveProperty("message");
+    expect(typeof deleteRes.message).toBe("string");
+    expect(deleteRes.message.length).toBeGreaterThan(0);
+  });
+
+  // The following feedback tests require valid org_id and project_id to be set in the environment variables (MEM0_ORG_ID, MEM0_PROJECT_ID).
```
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.integration.test.ts:76]: Can we not remove commented out code?

File: mem0-ts/src/client/tests/memoryClient.integration.test.ts [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    expect(typeof tempClient.telemetryId).toBe("string");
+  });
+
+  it("lists users", async () => {
+    const users = await client.users();
+    expect(users).toHaveProperty("results");
+    expect(Array.isArray(users.results)).toBe(true);
+  });
+
+  // The following tests for deleteUser, deleteUsers, and webhooks are included for completeness.
```
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.integration.test.ts:124]: Same as above

File: mem0-ts/src/client/tests/memoryClient.integration.test.ts
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.integration.test.ts:1]: Remove commented out code from the PR

File: mem0-ts/src/client/tests/memoryClient.test.ts [hunk 11]
```diff
@@ -1,391 +1,130 @@
...
-      expect(
-        Array.isArray(memory.categories) || memory.categories === null,
-      ).toBe(true);
-      if (Array.isArray(memory.categories)) {
-        memory.categories.forEach((category) => {
-          expect(typeof category).toBe("string");
-        });
+    it("updates a memory", async () => {
+      const updated: any = await client.update(memoryId, "Updated memory content");
+      if (Array.isArray(updated)) {
```
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.test.ts:53]: Why do we need a if/else here? Are we not sure what is the format of the response?

File: mem0-ts/src/client/tests/memoryClient.test.ts [hunk 12]
```diff
@@ -1,391 +1,130 @@
...
-        Array.isArray(memory.categories) || memory.categories === null,
-      ).toBe(true);
-      if (Array.isArray(memory.categories)) {
-        memory.categories.forEach((category) => {
-          expect(typeof category).toBe("string");
-        });
+    it("deletes the test user if present", async () => {
+      const allUsers = await client.users();
+      const entity = allUsers.results.find((u) => u.name === userId);
+      if (entity) {
```
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.test.ts:84]: Again why are we using conditionals????

File: mem0-ts/src/client/tests/memoryClient.test.ts [hunk 13]
```diff
@@ -1,391 +1,130 @@
...
-      if (Array.isArray(historyEntry.input)) {
-        historyEntry.input.forEach((input) => {
-          // Each input should be an object
-          expect(typeof input).toBe("object");
-
-          // Should have string content
-          expect(typeof input.content).toBe("string");
+    it("searches for a memory (v2)", async () => {
+      const results = await client.search("vegetarian", {
+        user_id: userId,
```
Comment (whysosaket) [inline mem0-ts/src/client/tests/memoryClient.test.ts:114]: Can we remove this?

File: mem0-ts/src/oss/tests/memory.advanced.test.ts
Comment (whysosaket) [inline mem0-ts/src/oss/tests/memory.advanced.test.ts:1]: Can we remove this file from the final PR?

**Comment by whysosaket** [comment]
@akshat1423 can you also make sure that we take care of the following,

1. Remove commented out code
2. Have a cleanup script that cleans the data for that particular user_id, agent_id etc
3. Attach screenshots of Tests Passing
4. Make sure that whenever you're doing a search/get not use something like [0] as tests run in parallel and can alter results later on.
5. Try to use V2 APIs as much as possible
